File: assignments/PPT/DSA/Assignments/Assignment_2/Q1.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <iterator>

using namespace std;

/*

💡 Given an integer array nums of 2n integers, group these integers into n pairs (a1, b1), (a2, b2),..., (an, bn) such that the sum of min(ai, bi) for all i is maximized. Return the maximized sum.

**Example 1:**
Input: nums = [1,4,3,2]
Output: 4

**Explanation:** All possible pairings (ignoring the ordering of elements) are:

1. (1, 4), (2, 3) -> min(1, 4) + min(2, 3) = 1 + 2 = 3
2. (1, 3), (2, 4) -> min(1, 3) + min(2, 4) = 1 + 2 = 3
3. (1, 2), (3, 4) -> min(1, 2) + min(3, 4) = 1 + 3 = 4
So the maximum possible sum is 4

*/

int solve(vector<int>& nums) {
    int sum = 0;
    sort(nums.begin(),nums.end());
    for(int i = 0;i<nums.size();i+=2){
        sum += nums[i];
    }
    return sum;
}

int main(){
    // Remove the following three lines when running in terminal with input or some interface
    #ifndef ONLINE_JUDGE
        freopen("input1.txt","r",stdin);
    #endif

    int t = 1;
    cin>>t;
    while(t--){
        int n,target;
        cin>>n;
        vector<int> ar(n);
        for(int i = 0;i<n;i++){
            cin>>ar[i];
        }

        int ans = solve(ar);
        cout<<ans<<endl;
    }

    return 0;
}

File: assignments/PPT/DSA/Assignments/Assignment_4/Q1.py
'''
Given three integer arrays arr1, arr2 and arr3 **sorted** in **strictly increasing** order, return a sorted array of **only** the integers that appeared in **all** three arrays.

**Example 1:**

Input: arr1 = [1,2,3,4,5], arr2 = [1,2,5,7,9], arr3 = [1,3,4,5,8]

Output: [1,5]

**Explanation:** Only 1 and 5 appeared in the three arrays.

'''

def solve(l1 : list, l2 : list, l3 : list) -> list:
    i = 0
    j = 0
    k = 0
    rl = list()

    while(i<len(l1) and j<len(l2) and k<len(l3)):
        if (l1[i] == l2[j] and l2[j] == l3[k]):
            rl.append(l1[i])
            i += 1
            j += 1
            k += 1
        elif l1[i] < l2[j]:
            i += 1
        elif l2[j] < l3[k]:
            j += 1
        else:
            k += 1
    return rl


if __name__ == "__main__":
    with open("./input1.txt",'r') as f:
        t = int(f.readline())
        for test_cases in range(t):
            l1 = [int(i) for i in f.readline().split(' ')]
            l2 = [int(i) for i in f.readline().split(' ')]
            l3 = [int(i) for i in f.readline().split(' ')]
            print(solve(l1,l2,l3))
  